[nullsafety] Non-nullable check warnings

RblSb opened the pull request:
Cannot undestand problem with warning diagnostics, need some help:
(nullsafety warnings will stay if there is nullsafety errors btw)
```haxe
@:nullSafety
@:haxe.warning("-WRedundantNullCheck")
class Main {
	static function main() {
		new Main();
	}

	function new() {
		final foo = "1";
		// this warning doesn't react on WRedundantNullCheck
		// and also hides after vscode tab swap
		final v = foo ?? "0";

		final foo = 1;
		// no warning because of Null<Int> cast in operator
		// probably can be fixed with some @:nonNullable internal meta on expr
		final v = foo ?? 0;

		// working warning example
		final f:Int;
		function name():Void {
			f + 1;
		}
	}
}
```

Closes https://github.com/HaxeFoundation/haxe/issues/7816

RblSb committed:
[nullsafety] Non-nullable check warnings

kLabz commented:
I will try to fix diagnostics vs cache early next week, that behavior is currently more or less expected.

Edit: was mostly about haxe.warning rules vs cached modules, but there are several things wrong in this PR that I'm working on fixing

kLabz commented:
Reworked to make it work with type level `@:haxe.warning` rules:
https://github.com/kLabz/haxe/commit/d9939f9e963bcc56b56907ec33bdfe5df43a7d32

[Kinda] works with:
```hx
@:nullSafety
@:haxe.warning("+WRedundantNullCheck") // Notice the "+" instead of "-"
class Main {
	static function main() {}

	// TODO
	// @:haxe.warning("+WRedundantNullCheck")
	function new() {
		final foo = "1";
		// this warning does react to +WRedundantNullCheck
		// but still hides after vscode tab swap (warnings added in filter happen too late.. need to think about it)
		final v = foo ?? "0";
		ignore(v);
	}

	static function ignore<T>(v:T):Void {}
}
```

Field/expression level is doable but more annoying to handle. Can definitely be done, though.
Will look into that warnings in filters vs cache thing

RblSb committed:
remove null wrap for null coalescing

RblSb committed:
Update kLabz commit

Use Common.module_warning, it doesn't have cache problem with compilation server / diagnostics

RblSb commented:
Field-level @:haxe.warning not implemented yet, but this can be reviewed, seems to work now

RblSb commented on tests/unit/src/unit/TestNullCoalescing.hx:
```diff
@@ -175,11 +177,13 @@ class TestNullCoalescing extends Test {
 		final a = nullString;
 		eq(a ?? "2", "2");
 
+		#if !static
 		eq(1 ?? 2, 1);
+		#end
```
Null-check doesn't work on static targets for basic types without Null<T> with this PR change now, so only dynamic targets are checked

Simn replied:
What he said. `a ?? b` expands to `a == null ? b : a`, and `1 == null` is not allowed on static targets. This is true for Int, Float and Bool values.

tobil4sk replied:
This is new behaviour right? We should add a misc test for it

Simn replied:
Is it? I thought that's how it always worked. But sure, more tests is always better!

tobil4sk replied:
It always behaved like that for `1 == null`, but I think failing for `a ?? b` is new (hence the new `#if static` checks)

RblSb commented on tests/unit/src/unit/TestNullCoalescing.hx:
```diff
@@ -175,11 +177,13 @@ class TestNullCoalescing extends Test {
 
+		#if !static
 		eq(1 ?? 2, 1);
+		#end
 		eq("1" ?? "2", "1");
```
Not a problem for strings on static tagets

Simn commented:
Weird, now when I try to ping @Copilot it just said `Copilot isn't available for cross-repository pull requests`. I guess I'll try to let it re-review.

RblSb committed:
Check binops in binops

RblSb commented on src/typing/nullSafety.ml:
```diff
@@ -6,16 +6,37 @@ open Error
+		sw_module = m;
+	} in
+	if not (List.mem warning report.sr_warnings) then
+		report.sr_warnings <- warning :: report.sr_warnings;
```
Duality of a semicolon

RblSb committed:
remove ;;

RblSb committed:
test failing ?? on cpp

yuxiaomao commented:
This PR also fixed the following regression on hashlink (according to tryhaxe, it works correctly on 4.3.7 but not with 5.0-preview1), and introduced https://github.com/HaxeFoundation/haxe/issues/12622 "errors" in Hide's CI.

Maybe it'll be good that I add a unit test on this one?

```haxe
enum abstract State(Int) {
	var Empty = 0;
	var Inquisition = 1;
}

function changeState(s:State) {
	trace(s);
}

var boardsForcedState:Null<State> = null;

function main() {
	changeState(boardsForcedState ?? boardsForcedState ?? Inquisition); // 0, expected 1
	var r = boardsForcedState ?? boardsForcedState ?? Inquisition; // normal
	changeState(r);
}
```

RblSb commented:
This is probably because i fixed `var a:String = null ?? ""` test case using `WithType.value` for `e1` typing instead of `with_type`, because temp var from `??` operator was failing in null safety while typed as nullable.

I think this is also good static runtime misc test to add.
